File: 6/average_of_levels_in_btree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def averageOfLevels(self, root: TreeNode) -> List[float]:
        if not root:
            return []
        
        row = [root]
        res = []
        while row:
            sum = 0
            nextrow = []
            for n in row:
                sum += n.val
                if n.left:
                    nextrow.append(n.left)
                if n.right:
                    nextrow.append(n.right)
            res.append(sum/len(row))
            row = nextrow
        return res

File: 6/average_of_levels_in_btree.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func averageOfLevels(root *TreeNode) []float64 {
	if root == nil {
		return make([]float64, 0)
	}

	res := make([]float64, 0)
	row := []*TreeNode{root}
	for len(row) > 0 {
		nextRow := make([]*TreeNode, 0)
		var sum float64
		for _, n := range row {
			sum += float64(n.Val)
			if n.Left != nil {
				nextRow = append(nextRow, n.Left)
			}
			if n.Right != nil {
				nextRow = append(nextRow, n.Right)
			}
		}
		res = append(res, sum/float64(len(row)))
		row = nextRow
	}
	return res
}

func main() {
	r := TreeNode{3,
		&TreeNode{9, nil, nil},
		&TreeNode{20,
			&TreeNode{15, nil, nil},
			&TreeNode{7, nil, nil}}}
	fmt.Println(averageOfLevels(&r))
}
